File: Test cases/kotlinLiterals.swift
//
// Copyright 2018 Vinicius Jorge Vendramini
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// gryphon output: Test cases/Bootstrap Outputs/kotlinLiterals.swiftAST
// gryphon output: Test cases/Bootstrap Outputs/kotlinLiterals.gryphonASTRaw
// gryphon output: Test cases/Bootstrap Outputs/kotlinLiterals.gryphonAST
// gryphon output: Test cases/Bootstrap Outputs/kotlinLiterals.kt

////////////////////////////////////////////////////////////////////////////////////////////////////

// Insert code at the beginning of the file
// gryphon insertInMain: println(\"Inserting at the beginning of the file\")

// Test imports in kotlin
// gryphon insert: import java.util.*
// gryphon insert:
// gryphon insert: fun myFunction(): String {
// gryphon insert: 	return \"Calling myFunction()\" // \\n \\t \\
// gryphon insert: }

////////////////////////////////////////////////////////////////////////////////////////////////////
import Foundation

// Kotlin literals as expressions
let languageName = "swift" // gryphon value: \"kotlin\"
print("Hello from \(languageName)!")

let magicNumber: Int = 40 + // gryphon value: 40 + 5-3
	2 // gryphon value: this comment will be ignored
print(magicNumber)

func f(a: Int = 0,
	   b: Int = 0 // gryphon value: 1
	) {
	print(a + b)
}

f(a: 0,
  b: 0 // gryphon value: 1
)

////////////////////////////////////////////////////////////////////////////////////////////////////
// Kotlin literals as statements
// gryphon insertInMain: println(\"This will be ignored by swift, but not by kotlin.\")
// gryphon insertInMain:
// gryphon insertInMain: println(myFunction())
// gryphon insertInMain: println(myOtherFunction())

////////////////////////////////////////////////////////////////////////////////////////////////////
// Ignore swift statements
print("This will be ignored by kotlin, but not by swift.") // gryphon ignore

////////////////////////////////////////////////////////////////////////////////////////////////////
// Call something swift can't parse
let squareRoot = sqrt(9) // gryphon value: Math.sqrt(9.0)
print(squareRoot)

////////////////////////////////////////////////////////////////////////////////////////////////////
// Ignore a swift declaration
class IgnoredClass // gryphon ignore
{
}
enum IgnoredEnum // gryphon ignore
{
}
enum IgnoredStruct // gryphon ignore
{
}

func a() { // gryphon ignore
}

////////////////////////////////////////////////////////////////////////////////////////////////////
// Test annotations
protocol A {
	var x: Int { get }
	var y: Int { get }

	func foo()
}

class B: A {
	var x: Int = 1 // gryphon annotation: override
	var y: Int = // gryphon annotation: override
		2 // gryphon value: 3
	var z: Int = 0

	func foo() { } // gryphon annotation: override
}

print(B().x)
print(B().y)

protocol C { // gryphon annotation: open
}

enum D { // gryphon annotation: final
	case a
}

struct E { // gryphon annotation: final
	let a: Int = 0
}

////////////////////////////////////////////////////////////////////////////////////////////////////
// Test inserts and declarations
if true {
	// gryphon insert: println(\"if 1\")
	print("if 2")
	// gryphon insert: println(\"if 3\")
	print("if 4")
	// gryphon insert: println(\"if 5\")
}

for i in [1] {
	// gryphon insert: println(\"for 1\")
	print("for 2")
	// gryphon insert: println(\"for 3\")
	print("for 4")
	// gryphon insert: println(\"for 5\")
}

func insertFunction() {
	// gryphon insert: println(\"func 1\")
	print("func 2")
	// gryphon insert: println(\"func 3\")
	print("func 4")
	// gryphon insert: println(\"func 5\")
}
insertFunction()

class InsertClass {
	// gryphon insert: open var a: String = \"class a\"
	var b: String = "class b"
	// gryphon insert: open var c: String = \"class c\"
	var d: String = "class d"
	// gryphon insert: open var e: String = \"class e\"

	func insertMethod() {
		// gryphon insert: println(\"method 1\")
		print("method 2")
		// gryphon insert: println(\"method 3\")
		print("method 4")
		// gryphon insert: println(\"method 5\")
	}
}
let insertClass = InsertClass()
// gryphon insertInMain: println(insertClass.a)
// gryphon insertInMain: println(insertClass.b)
// gryphon insertInMain: println(insertClass.c)
// gryphon insertInMain: println(insertClass.d)
// gryphon insertInMain: println(insertClass.e)
insertClass.insertMethod()


////////////////////////////////////////////////////////////////////////////////////////////////////
// Insert code at the end of the file

// gryphon insertInMain: println(\"Code at the end of file.\")

// gryphon insert: fun myOtherFunction(): String {
// gryphon insert: 	return \"Calling myOtherFunction()\" // \\n \\t \\
// gryphon insert: }

File: Test cases/kotlinLiterals.kt
//
// Copyright 2018 Vinicius Jorge Vendramini
//
// Licensed under the Hippocratic License, Version 2.1;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://firstdonoharm.dev/version/2/1/license
//
// To the full extent allowed by law, this software comes "AS IS,"
// WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED, and licensor and any other
// contributor shall not be liable to anyone for any damages or other
// liability arising from, out of, or in connection with the sotfware
// or this license, under any kind of legal claim.
// See the License for the specific language governing permissions and
// limitations under the License.
//
import java.util.*

fun myFunction(): String {
	return "Calling myFunction()" // \n \t \
}

internal fun f(a: Int = 0, b: Int = 1) {
	println(a + b)
}

internal interface A {
	val x: Int
	val y: Int

	fun foo()
}

internal open class B: A {
	override open var x: Int = 1
	override open var y: Int = 3
	open var z: Int = 0

	override open fun foo() {
	}
}

open internal interface C {
}

final internal enum class D {
	A;
}

final internal data class E(
	val a: Int = 0
)

internal fun insertFunction() {
	println("func 1")

	println("func 2")

	println("func 3")

	println("func 4")

	println("func 5")
}

internal open class InsertClass {
	open var a: String = "class a"

	open var b: String = "class b"

	open var c: String = "class c"

	open var d: String = "class d"

	open var e: String = "class e"

	open fun insertMethod() {
		println("method 1")

		println("method 2")

		println("method 3")

		println("method 4")

		println("method 5")
	}
}

fun myOtherFunction(): String {
	return "Calling myOtherFunction()" // \n \t \
}

fun main(args: Array<String>) {
	println("Inserting at the beginning of the file")

	val languageName: String = "kotlin"

	println("Hello from ${languageName}!")

	val magicNumber: Int = 40 + 5-3

	println(magicNumber)
	f(a = 0, b = 1)

	println("This will be ignored by swift, but not by kotlin.")

	println(myFunction())
	println(myOtherFunction())

	val squareRoot: Double = Math.sqrt(9.0)

	println(squareRoot)
	println(B().x)
	println(B().y)

	if (true) {
		println("if 1")

		println("if 2")

		println("if 3")

		println("if 4")

		println("if 5")
	}

	for (i in listOf(1)) {
		println("for 1")

		println("for 2")

		println("for 3")

		println("for 4")

		println("for 5")
	}

	insertFunction()

	var closure: () -> Unit = {
			println("closure 1")

			println("closure 2")

			println("closure 3")

			println("closure 4")

			println("closure 5")
		}

	closure()

	val insertClass: InsertClass = InsertClass()

	println(insertClass.a)
	println(insertClass.b)
	println(insertClass.c)
	println(insertClass.d)
	println(insertClass.e)

	insertClass.insertMethod()

	println("Code at the end of file.")
}
